respect explicit return values from controller functions (fixes #11147)

jamestalmage opened the pull request:
When controller functions return an explicit value that value should be what is passed to the linking functions, and to any child/sibling controllers that `require` it. It should also be bound to the data store on the DOM element.

Closes #11147

jamestalmage committed:
test(compile): failing test - respect controller return value

failing test for angular/angular.js#11147

Controllers instantiated for directives by the $compile service
do not respect explicit return values from the controller function.

jamestalmage committed:
fix(compile): respect explicit return values from controller functions

When controller functions return an explicit value that value should
be what is passed to the linking functions, and to any child/sibling
controllers that `require` it. It should also be bound to the data
store on the dom element.

Closes #11147

jamestalmage commented:
CLA updated

petebacondarwin commented:
@caitp you have done some work on this part of the code. Can you take a look at this one?

caitp commented:
the only thing that's needed here is updating the element data

caitp commented on src/ng/compile.js:
```diff
@@ -1967,13 +1967,16 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-              // Remove and re-install bindToController bindings
-              thisLinkFn.$$destroyBindings();
-              thisLinkFn.$$destroyBindings =
+            if (controllerResult !== controller.instance && (isObject(controllerResult) || isFunction(controllerResult))) {
```
in particular, don't worry about the `isObject() || isFunction()` stuff, since that's handled in controller.js anyways

jamestalmage replied:
I'm not sure what you are asking for here. We should mimic the behavior of native javascript: If they return an object or function use that as the result of the instantiation, otherwise return "as normal":

```javascript
function returnsStr() {
  this.foo = "bar";
  return "baz";
}

new returnsStr() === {foo:"bar"} // "baz" return value is ignored
```

caitp replied:
this code is duplicating code which already runs in the `controller()` function itself, it's not needed here

caitp replied:
particularly [here](https://github.com/angular/angular.js/blob/62d514b06937cc7dd86e973ea11165c88343b42d/src/ng/controller.js#L139-L146)

31812015 replied:
this diff

jamestalmage replied:
Nevermind,
I see what you are saying, my above concern is already handled by the call to the `controller()` callback. I am adding a test case that verifies it.

jamestalmage replied:
argh!
Come on Github! Real time updates for line comments!

caitp commented on src/ng/compile.js:
```diff
@@ -1967,13 +1967,16 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
-              thisLinkFn.$$destroyBindings =
+            if (controllerResult !== controller.instance && (isObject(controllerResult) || isFunction(controllerResult))) {
+              controller.instance = controllerResult;
+              $element.data('$' + directive.name + 'Controller', controllerResult);
```
there are a few different circumstances that this happens, so we might need to be a bit more careful about this --- be sure to include test coverage that covers controllers for transclude directives too

caitp commented on src/ng/compile.js:
```diff
@@ -1967,13 +1967,16 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+                thisLinkFn.$$destroyBindings =
                   initializeDirectiveBindings(scope, attrs, controllerResult,
-                                              bindings, scopeDirective);
+                    bindings, scopeDirective);
```
don't change indentation

caitp commented:
LGTM with comments addressed

jamestalmage commented:
unclear what this means

> the only thing that's needed here is updating the element data

Are you asking me to delete the first of [these two lines](https://github.com/jamestalmage/angular.js/commit/57fe7014c7187cb796af4d483e913c3a0dd29d38#diff-a732922b631efed1b9f33a24082ae0dbR1971)?:
```javascript
controller.instance = controllerResult;
$element.data('$' + directive.name + 'Controller', controllerResult);
```
Both lines are required for my test to pass.

caitp commented:
no, you need that line as mentioned above --- I would like you to add additional tests, though.

1. add another test where a controller which returns a custom object is require'd by another directive, assert that the directive gets the right controller (this is sort of what you're testing already, but it tests what we actually want, rather than the mechanics of it)
2. add similar test cases which verify the transclude cases

jamestalmage commented:
> `2. add similar test cases which verify the transclude cases

your use of plural case(**s**) is sending me down a rabbit hole in `compileSpec`. There are many tests related to transclusion and controllers. 

I've added this one so far:
```javascript
    it('should support controllerAs', function() {
      module(function() {
        directive('main', function() {
          return {
            templateUrl: 'main.html',
            transclude: true,
            scope: {},
            controller: function() {
              this.name = 'lucas';
              return {name: 'george'};
            },
            controllerAs: 'mainCtrl'
          };
        });
      });
      inject(function($templateCache, $compile, $rootScope) {
        $templateCache.put('main.html', '<span>template:{{mainCtrl.name}} <div ng-transclude></div></span>');
        element = $compile('<div main>transclude:{{mainCtrl.name}}</div>')($rootScope);
        $rootScope.$apply();
        expect(element.text()).toBe('template:george transclude:');
      });
    });
```

But there are plenty of other cases. You want me to go nuts and add dozens of test cases, or are there specific scenarios you want tested.

caitp commented:
you don't need "dozens" of tests, but add ones that deal with a directive whose controller gets the transclude function

jamestalmage committed:
test(compile): additional tests for controllers with explicit return values.

Includes tests for controllers that:
 - explicitly return primitives
 - are attached to scope using `controllerAs`
 - transclude contents

jamestalmage committed:
refactor(compile): remove duplicate type checks

The `controller()` callback already performs the required type
checks and guarantees that `controller() !== controller.instance`
in the event that the function returns an explicit value.

jamestalmage commented:
@caitp  I believe I have addressed all of your concerns, but I am not familiar enough with the inner workings of transclude to be sure I have covered all cases. Please let me know if there is a scenario I have missed (pointers to existing tests would be helpful).

caitp commented on test/ng/compileSpec.js:
```diff
@@ -5525,6 +5615,41 @@ describe('$compile', function() {
+              return {transclude:$transclude, foo: 'bar'};
+            },
+            link: function(scope, el, attr, ctrl) {
+              ctrl.transclude(cloneAttach);
```
we don't need to actually run the transclude fn, we just need to assert that it's there

caitp replied:
It would be good to make these test cases a little closer, so that the only difference between them is the presence of the transclude property of the DDO. almost there =)

jamestalmage replied:
so proving each clone gets the correct object reference to the parent controller has no value? (serious question)

caitp replied:
no, this is tested elsewhere --- we only really care that the nested directive gets the right controller

jamestalmage replied:
in my mind
`"each clone gets the correct object reference to the parent controller" === "the nested directive gets the right controller"`.

I need to call `$transclude` at least once to see what the nested controller gets, right?

caitp replied:
well to link the nested directive yes, but you don't need to call it 3 times --- an `expect(controller).toBe(the right thing)` is the only real assertion that is needed

jamestalmage replied:
understood.
on it.

jamestalmage committed:
test(compile): apply discussed changes from #11326

jamestalmage commented:
@caitp, made most of the changes requested, except this one:

> It would be good to make these test cases a little closer, so that the only difference between them is the presence of the transclude property of the DDO

Can you clarify which tests you want to look closer?

caitp commented:
merged, thanks for the work

jamestalmage commented:
awesome. Thank you!
